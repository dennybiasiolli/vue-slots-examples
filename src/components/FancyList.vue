<script setup lang="ts">
import { onMounted, ref } from 'vue'

export interface Props {
  apiUrl: string
}

defineProps<Props>()

interface Item {
  id: number
  body: string
  username: string
  likes: number
}

const items = ref<Item[]>([])

function getExternalData(): Promise<Item[]> {
  return new Promise((resolve) =>
    resolve([
      { id: 1, body: "Cyborgs don't feel pain.", username: 'kyle_reese', likes: 3 },
      { id: 2, body: 'Can you stop it?', username: 'sarah_connor', likes: 7 },
      { id: 3, body: 'Who sent you?', username: 'john_connor', likes: 5 }
    ])
  )
}

onMounted(async () => {
  items.value = await getExternalData()
})
</script>

<template>
  <span>Calling {{ apiUrl }}</span>
  <ul>
    <li v-for="item in items" :key="item.id">
      <slot name="item" v-bind="item">{{ item.body }}</slot>
    </li>
  </ul>
</template>
