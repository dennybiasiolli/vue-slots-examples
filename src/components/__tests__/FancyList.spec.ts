import { describe, it, expect } from 'vitest'

import { mount } from '@vue/test-utils'
import FancyList from '../FancyList.vue'

describe('FancyList', () => {
  it('with default content', async () => {
    const wrapper = mount(FancyList, { props: { apiUrl: 'foo/bar/' } })
    await wrapper.vm.$nextTick()
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('with item slot', async () => {
    const wrapper = mount(FancyList, {
      props: { apiUrl: 'foo/bar/' },
      slots: {
        item: `Custom render: {{ params.body }}`
      }
    })
    await wrapper.vm.$nextTick()
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })
})
