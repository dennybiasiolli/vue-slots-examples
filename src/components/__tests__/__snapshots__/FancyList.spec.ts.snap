// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FancyList > with default content 1`] = `
"<span>Calling foo/bar/</span>
<ul>
  <li>Cyborgs don't feel pain.</li>
  <li>Can you stop it?</li>
  <li>Who sent you?</li>
</ul>"
`;

exports[`FancyList > with item slot 1`] = `
"<span>Calling foo/bar/</span>
<ul>
  <li>Custom render: Cyborgs don't feel pain.</li>
  <li>Custom render: Can you stop it?</li>
  <li>Custom render: Who sent you?</li>
</ul>"
`;
